embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt

```diff
@@ -102,33 +116,49 @@ internal class EmbraceDeliveryCacheManager(
         }
     }
 
+    /**
+     * This method will do disk reads so do not run it on the main thread
+     */
     override fun getAllCachedSessionIds(): List<String> {
         val allSessions = cacheService.listFilenamesByPrefix(SESSION_FILE_PREFIX)
         allSessions.forEach { filename ->
             if (filename == OLD_VERSION_FILE_NAME) {
```

Comment:
So we can only ever have one cached session from the old SDK version? In the current SDK, is that basically true as well, but we just have a special timestamp in the file name?

Reply:
From what I remember older versions of the SDK wrote sessions to a single filename, and then on launch converted that file into a session that can be sent to the server. This was not good as it meant we needed to perform I/O at startup to avoid data loss, so we added a timestamp + UUID to the filename to get around this. So yes, there should only be 1 cached session from older SDK versions.

@lucaslabari your knowledge is probably better than mine here, please correct me if I got anything wrong

Reply:
Well there's another vector for losing sessions 😅 

BTW, I didn't make the change here, but I think we should defer this until app startup is complete (or do it in the background at least). It's a manner of less than a couple seconds in most cases, but doing this can be quite expensive

Reply:
I agree that'd be sensible as long as the previous process didn't crash, as we don't want to impact deliverability if we're in the middle of a crash loop

Reply:
IMO the higher probability is worth it. If the SDK flushes sessions at different points of startup depending on what happened last process, it also makes it easier to measure the production perf impact of initializing later

Reply:
> From what I remember older versions of the SDK wrote sessions to a single filename, and then on launch converted that file into a session that can be sent to the server. This was not good as it meant we needed to perform I/O at startup to avoid data loss, so we added a timestamp + UUID to the filename to get around this. So yes, there should only be 1 cached session from older SDK versions.

@fractalwrench yeap, you are correct. Apart from performing I/O at startup, the problem with the older approach was that the cache of a session could override a previous one in cases where for example there is a new crash at startup before the previous session was intended to be sent, so it wasn't sent nor saved as a pending API call. Resulting in a session being lost. 
I think we can defer the sent of the session if we want to improve startup time and in case we want to differentiate the behavior in the cases where the previous session crashed, we can use the `crash_marker_file` to know that and do it at an early stage if that file is present.

Reply:
We can address all of this when we do that work. The native stuff is probably worth deferring more, so we can look at this holistically. I'll create a ticket to track this more formally.

I don't mind startup taking longer if we actually have a session to send - I think for me, we are paying the hit regardless if there's a cached session, so even doing a quick no-op would address this satisfactorily for me.